### PR TITLE
Integrate the Singularity image build process into the Travis CI and move Travis CI to run on Ubuntu Xenial

### PR DESCRIPTION
**Pull request by tai271828**
Ubuntu Trusty is going to end-of-life few months later. Let's move to Xenial. Besides, build a Singularity image on this Xenial-based flow.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -57,6 +64,24 @@ before_install:
         $DOCKER)
 
       SCRIPT_RUN_PREFIX="docker exec --tty $containerid"
+    elif [[ "$RUNMODE" == "singularity" ]] ; then
+      # prepare Singularity
```
Comment (yungyuc) [inline .travis.yml:68]: How much time does this build step take?  I don't want it to too long.

File: .travis.yml [hunk 2]
```diff
@@ -89,45 +114,70 @@ install:
...
+    if [[ "$RUNMODE" != "docker" ]] && [[ "$RUNMODE" != "singularity" ]]; then
       # Install conda packages
       ${TRAVIS_BUILD_DIR}/contrib/devenv/create.sh
       source ${TRAVIS_BUILD_DIR}/build/env/start
       ${TRAVIS_BUILD_DIR}/contrib/conda.sh
       ${TRAVIS_BUILD_DIR}/contrib/build-pybind11-in-conda.sh
       # Debugging information
       conda info -a
-    else
+    elif [[ "$RUNMODE" == "docker"  ]]; then
```
Comment (yungyuc) [inline .travis.yml:125]: A redundant space between `"` and `]]`.

File: .travis.yml (hunk 1 above)
Comment (yungyuc) [reply .travis.yml:68]: If singularity has a Ubuntu package we probably should use it to save install time?

File: .travis.yml [hunk 3]
```diff
@@ -89,45 +114,70 @@ install:
...
       $SCRIPT_RUN_PREFIX apt-get -qqy install fakeroot debhelper locales \
         libreadline7 libssl1.0.0 libffi6 \
         liblapack3 liblapack-dev libhdf5-100 libhdf5-dev libnetcdf13 \
         libnetcdf-dev libscotch-6.0 libscotch-dev cython3 python3 \
         python3-numpy libpython3.6-dev python3-boto python3-paramiko graphviz
+    elif [[ "$RUNMODE" == "singularity"  ]]; then
+      pushd ${SHOME}
+      cp ${TRAVIS_BUILD_DIR}/sandbox/gas/tube/${SOLVCON_DRIVING} ${SHOME}
+      # trick to inject TRAVIS_BRANCH and TRAVIS_REPO_SLUG
+      echo '#!/bin/bash' > ${SHOME}/prepare-solvcon-dev.sh
```
Comment (yungyuc) [inline .travis.yml:137]: I think it's better to use

```bash
cat << EOF
...
EOF
```

File: .travis.yml [hunk 4]
```diff
@@ -89,45 +114,70 @@ install:
...
+      echo "export TRAVIS_BRANCH=${TRAVIS_BRANCH}" >> ${SHOME}/prepare-solvcon-dev.sh
+      echo "export TRAVIS_REPO_SLUG=${TRAVIS_REPO_SLUG}" >> ${SHOME}/prepare-solvcon-dev.sh
+      cat ${TRAVIS_BUILD_DIR}/contrib/singularity/prepare-solvcon-dev.sh >> ${SHOME}/prepare-solvcon-dev.sh
+      cp ${TRAVIS_BUILD_DIR}/contrib/singularity/${SRECIPE} ${SHOME}
+      sudo singularity build ${SIMAGE} ${SRECIPE}
```
Comment (yungyuc) [inline .travis.yml:142]: I remember that I wanted to test singularity image building in travis.  However, since we aren't building the docker image here, we probably shouldn't do it for singularity, either.

I still think we should build (and test) the singularity image automatically.  But mixing the CI for SOLVCON and its containers here looks weird, after I see the code now.  And it makes CI slow.

Will there be a better way to handle both testing?

File: .travis.yml [hunk 5]
```diff
@@ -89,45 +114,70 @@ install:
...
   # Debugging information
   - |
-    $SCRIPT_RUN_PREFIX which python3
-    $SCRIPT_RUN_PREFIX python3 --version
-    $SCRIPT_RUN_PREFIX which $CXX
-    $SCRIPT_RUN_PREFIX $CXX --version
-    $SCRIPT_RUN_PREFIX which $CC
-    $SCRIPT_RUN_PREFIX $CC --version
-  - $SCRIPT_RUN_PREFIX make SC_PURE_PYTHON=1 test_from_package
+    # We do everything in the singularity container including unit tests. The
```
Comment (yungyuc) [inline .travis.yml:150]: We don't need to change this.  Having the information printed isn't harmful.

File: .travis.yml [hunk 6]
```diff
@@ -89,45 +114,70 @@ install:
...
+      $SCRIPT_RUN_PREFIX which python3
+      $SCRIPT_RUN_PREFIX python3 --version
+      $SCRIPT_RUN_PREFIX which $CXX
+      $SCRIPT_RUN_PREFIX $CXX --version
+      $SCRIPT_RUN_PREFIX which $CC
+      $SCRIPT_RUN_PREFIX $CC --version
+    fi
+  - |
+    if [[ "$RUNMODE" != "singularity" ]] ; then
+      $SCRIPT_RUN_PREFIX make SC_PURE_PYTHON=1 test_from_package
```
Comment (yungyuc) [inline .travis.yml:164]: This step should be merged to the next one in the `elif`.

File: .travis.yml [hunk 7]
```diff
@@ -89,45 +114,70 @@ install:
...
+    elif [[ "$RUNMODE" == "singularity" ]] ; then
+      $SCRIPT_RUN_PREFIX ${TRAVIS_BUILD_DIR}/sandbox/gas/tube/${SOLVCON_DRIVING} run tube
     fi
 
 after_script:
   - |
     if [[ "$RUNMODE" == "docker" ]] ; then
       docker stop "$containerid"
       docker rm "$containerid"
+    elif [[ "$RUNMODE" == "singularity" ]] ; then
```
Comment (yungyuc) [inline .travis.yml:181]: This is another reason that I don't like to build singularity from source right here.

File: contrib/singularity/prepare-solvcon-dev.sh [hunk 8]
```diff
@@ -15,7 +15,17 @@ export PATH="${SCSRC}:${MINICONDA_DIR}/bin/:${PATH}"
 mkdir -p ${SCSRC_WORKING}
 
 # fetch SOLVCON source
-git clone https://github.com/solvcon/solvcon.git ${SCSRC}
+SCSRC_URL="https://github.com/solvcon/solvcon.git"
```
Comment (yungyuc) [inline contrib/singularity/prepare-solvcon-dev.sh:18]: How about:

```bash
SCSRC_URL=${SCSRC_URL:-https://github.com/solvcon/solvcon.git}
```

**Review by yungyuc — CHANGES_REQUESTED** [review]
We need to bring down the time for the singularity run.  It takes 12 minutes, twice the time of other runs.

File: .travis.yml (hunk 1 above)
Comment (tai271828) [reply .travis.yml:68]: There is no deb package for Ubuntu so far. It takes so much time because it builds from scratch, including an Ubuntu image. In my opinion, we could prepare a cache of the base image to reduce the build time.

File: .travis.yml (hunk 4 above)
Comment (tai271828) [reply .travis.yml:142]: My gut feeling told me we should leverage Singularity Hub, which is aim at building Singularity image. The image built by Singularity Hub could be pulled to somewhere so the travis CI could cache it and focus on running SOLVCON driving script. I need to evaluate if it is feasible.

Overall, I incline to the idea that we should not build the associated Singularity image here.

File: .travis.yml (hunk 5 above)
Comment (tai271828) [reply .travis.yml:150]: `$SCRIPT_RUN_PREFIX python3 --version` will make the return code 1 if the build env does not install conda. Regarding Singularity, we don't need to install conda outside the container.

File: .travis.yml (hunk 7 above)
Comment (tai271828) [reply .travis.yml:181]: We could probably build our deb and store it somewhere (launpad PPA for example) and install Singularity deb. The we could merge the singularity mode as a part of the regular test.

How to build the deb? Please refer to [Build a DEB from source of Singularity installation doc
](https://singularity.lbl.gov/install-linux). I will take care of this.

File: .travis.yml (hunk 1 above)
Comment (yungyuc) [reply .travis.yml:68]: That is better than building it on the fly in this travis file.

File: .travis.yml (hunk 4 above)
Comment (yungyuc) [reply .travis.yml:142]: If Singularity Hub hosts our image, in theory we can pull it in travis vm just like how we do it with the docker image.

File: .travis.yml (hunk 5 above)
Comment (yungyuc) [reply .travis.yml:150]: I'd prefer the debugging information be printed for all runs.

And If `python3` isn't in the path, there's a bug.  The executable should be always available in the path.

**Comment by tai271828** [comment]
tl;dr this pull request is oudated. We can make another one when we need it.

By the way, Singularity is renamed Apptainer, and moved into the Linux Foundation!! The original name "Singularity" will be used for commercial product line owned by Sylabs.

https://apptainer.org/news/community-announcement-20211130/
